mlir/lib/ExecutionEngine/conv-validation-wrappers.cpp

```diff
@@ -842,3 +843,188 @@ mcpuConv2dInt8(int64_t rank1, void *f_ptr, int64_t rank2, void *i_ptr,
       stride_h, stride_w, padding_h_l, padding_h_r, padding_w_l, padding_w_r,
       dilation_h, dilation_w, xdlops);
 }
+
+int findIdxHistRelDiff(double relDiff) {
+  /*
+    hist_relDiff[]:
+    0: 0
+    1: 0 - 1e-6
+    2: 1e-6 - 1e-5
+    3: 1e-5 - 1e-4
+    4: 1e-4 - 1e-3
+    5: 1e-3 - 1e-2
+    6: 1e-2 - 0.1
+    7: 0.1 - 1
+    8: >= 1
+   */
+  if (relDiff == 0.0)
+    return 0;
+  if (relDiff < 1.0e-06)
+    return 1;
+  if (relDiff < 1.0e-05)
+    return 2;
+  if (relDiff < 1.0e-04)
+    return 3;
+  if (relDiff < 1.0e-03)
+    return 4;
+  if (relDiff < 1.0e-02)
+    return 5;
+  if (relDiff < 0.1)
+    return 6;
+  if (relDiff < 1.0)
+    return 7;
+  return 8;
+}
+
+template <typename T>
+void mcpuVerify(T *gpuResults, T *validationResults, int64_t dataSize,
+                float thr_RMS, float thr_absDiff, float thr_relDiff) {
+  float valNum, gpuNum;
+  // metric maxAbsDiff
+  float maxAbsDiff = 0.0f;
+  double sumAbsDiff = 0.0;
+  float maxVAL_abs = 0.0f;
+  float maxGPU_abs = 0.0f;
+  // metric maxRelDiff
+  double maxRelDiff = 0.0;
+  double sumRelDiff = 0.0;
+  float maxVAL_rel = 0.0f;
+  float maxGPU_rel = 0.0f;
+  // Metric RMS
+  float maxMag = 0.0f;
+  double sumDiffSq = 0.0;
+  // histogram
+  // hist_relDiff[9]++ when cpuVal == 0
+  int hist_relDiff[10] = {0};
+  for (int64_t i = 0; i < dataSize; ++i) {
```

Comment:
Could we extend this loop sot hat, if a value goes over the absolute or relative threshold, we print the index of the value, the expected result, and the actual result?